packages/migrations/src/actions/2023.01.17T10.46.28.import-legacy-s3-keys-to-database.mts

```diff
@@ -0,0 +1,613 @@
+  S3_SECRET_ACCESS_KEY: zod.string().min(1),
+  S3_ENDPOINT: zod.string().url(),
+  S3_BUCKET_NAME: zod.string().min(1),
+  ENCRYPTION_SECRET: zod.string().min(1),
+});
+
+type Cursor = {
+  lastId: string;
+  lastCreatedAt: string;
+};
+
+export const up: slonik.Migration = async ({ context: { connection, sql } }) => {
+  // eslint-disable-next-line no-process-env
+  const eenv = shouldRunModel.parse(process.env);
+  const shouldRun = eenv.RUN_S3_LEGACY_CDN_KEY_IMPORT === '1';
+
+  if (!shouldRun) {
+    return;
+  }
+
+  // eslint-disable-next-line no-process-env
+  const env = envModel.parse(process.env);
+
+  const s3Client = new AwsClient({
+    accessKeyId: env.S3_ACCESS_KEY_ID,
+    secretAccessKey: env.S3_SECRET_ACCESS_KEY,
+    service: 's3',
+  });
+
+  function generateLegacyCDNAccessToken(targetId: string): string {
+    const encoder = new TextEncoder();
+    return createHmac('sha256', env.ENCRYPTION_SECRET)
+      .update(encoder.encode(targetId))
+      .digest('base64');
+  }
+
+  async function getPaginationTargets(
+    cursor: null | Cursor,
+  ): Promise<zod.TypeOf<typeof TargetsModel>> {
+    // Note: there is no index for this query, so it might be slow
+    // Also all this code runs inside a database transaction.
+    // This will block any other writes to the table.
+    // As the table should not be heavily in use when this is being run, it does not really matter.
+    const query = sql`
+      SELECT
+        "id"
+        , to_json("created_at") as "created_at_cursor"
+      FROM
+        "targets"
+      ${
+        cursor
+          ? sql`
+              WHERE
+                ("created_at" = ${cursor.lastCreatedAt} AND "id" > ${cursor.lastId})
+                OR "created_at" > ${cursor.lastCreatedAt}
+            `
+          : sql``
+      }
+      ORDER BY
+        "created_at" ASC
+        , "id" ASC
+      LIMIT
+        20
+    `;
+
+    const items = await connection.query(query);
+    return TargetsModel.parse(items.rows);
+  }
+
+  let lastCursor: null | Cursor = null;
+
+  do {
+    const items: zod.TypeOf<typeof TargetsModel> = await getPaginationTargets(lastCursor);
+    for (const item of items) {
+      const s3Key = `s3-legacy-keys/${item.id}`;
+      const privateAccessKey = generateLegacyCDNAccessToken(item.id);
+      const firstCharacters = privateAccessKey.substring(0, 3);
+      const lastCharacters = privateAccessKey.substring(privateAccessKey.length - 3);
+      const accessKeyHash = await bcryptjs.hash(privateAccessKey, await bcryptjs.genSalt());
+
+      // After creation on database
+      const response = await s3Client.fetch(
+        [env.S3_ENDPOINT, env.S3_BUCKET_NAME, s3Key].join('/'),
+        {
+          method: 'PUT',
+          body: accessKeyHash,
+        },
+      );
+
+      // Should probably do a retry here
+      if (response.status !== 200) {
+        throw new Error(`Unexpected Status for storing key. (status=${response.status})`);
```

Comment:
seems like the rest of the migration script is kinda graceful if it runs twice, any reason to throw an error error and just stop the migration from running?